contracts/adapters/CharmAdapter.sol

```diff
@@ -0,0 +1,388 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.7.2;
+pragma experimental ABIEncoderV2;
+
+import {
+    AggregatorV3Interface
+} from "@chainlink/contracts/src/v0.6/interfaces/AggregatorV3Interface.sol";
+import {SafeMath} from "@openzeppelin/contracts/math/SafeMath.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import {IProtocolAdapter, ProtocolAdapterTypes} from "./IProtocolAdapter.sol";
+import {
+    IOptionFactory,
+    IOptionMarket,
+    IOptionToken,
+    IOptionViews
+} from "../interfaces/CharmInterface.sol";
+import {IWETH} from "../interfaces/IWETH.sol";
+import {UniERC20} from "../lib/UniERC20.sol";
+
+contract CharmAdapter is IProtocolAdapter{
+    using SafeMath for uint256;
+    using SafeERC20 for IERC20;
+    using UniERC20 for IERC20;
+
+    struct OptionType {
+        bool isLongToken;
+        uint256 strikeIndex;
+    }
+
+    IOptionFactory public immutable optionFactory;
+    IOptionViews public immutable optionViews;
+
+    string private constant _name = "CHARM";
+    bool private constant _nonFungible = false;
+    mapping(bytes32 => address) private idToAddress;
+    mapping(address => OptionType) private addressToOptionType;
```

Comment:
So dealing with storage for adapters is tricky. Since an instrument talks to Adapters with delegatecall, it means that any reads/writes to storage will be done on the instrument level.

Here what we'll have to do is to add a [new version to the InstrumentStorage](https://github.com/ribbon-finance/structured-products/blob/master/contracts/storage/InstrumentStorage.sol#L37).

```solidity
contract InstrumentStorageV3 {
  mapping(bytes32 => address) private idToAddress;
  mapping(address => OptionType) private addressToOptionType;
}

// then you have to tack on the new storage to the instrument
contract RibbonVolatility is DSMath, InstrumentStorageV1, InstrumentStorageV2, InstrumentStorageV3 {
```

You can read more about this pattern of [writing upgradeable contracts here](https://docs.openzeppelin.com/upgrades-plugins/1.x/writing-upgradeable#modifying-your-contracts).